web3/utils/validation.py

```diff
@@ -33,11 +36,23 @@ def validate_abi(abi):
     """
     Helper function for validating an ABI
     """
+    seen_selectors = dict()
     if not is_list_like(abi):
         raise ValueError("'abi' is not a list")
     for e in abi:
         if not is_dict(e):
             raise ValueError("The elements of 'abi' are not all dictionaries")
+        if e['type'] == 'function':
+            selector = encode_hex(function_abi_to_4byte_selector(e))
+            if selector in seen_selectors:
```

Comment:
A small improvement to this would be to find **all** collisions and then raise.  I've never established a clean way to do this trivially, but something like the following works.

```python
selector_data = collections.defaultdict(list)
for e in abi:
    selector = ...
    selector_data[selector].append(signature)
collisions = {
    selector: signatures
    for selector, signatures
    in selector_data.items()
    if len(signatures) > 1
}
... # raise exception.
```

Reply:
Here's what I was thinking for a `groupby` approach:

```py
functions = [el for el in abi if el['type'] == 'function']
selectors = groupby(function_abi_to_4byte_selector, functions)
duplicates = valfilter(lambda funcs: len(funcs) > 1, selectors)
if len(duplicates) > 1:
    raise ...
```

---

The dict check could be:

```py
if not all(is_dict(el) for el in abi):
    raise ...
```

That way we can completely get rid of the `for e in abi` outer loop.

Reply:
```python
# instead of
functions = [el for el in abi if el['type'] == 'function']
#
functions = web3.utils.abi.filter_by_type('function', abi)
```

Reply:
This whole approach seems so wasteful to me. I'm not a big fan of wasting a for loop.
```python
[el for el in abi if el['type'] == 'function']
all(is_dict(el) for el in abi)
```
 I don't understand what advantage the code snippets above offers over the code that I've written. Although, spitting out all function collisions is missing, it can be easily implemented using a `defaultdict(list)` as piper suggested.
In fact the `groupby` wastes another loop which again can be avoided using `defaultdict(list)`.

IMO, using functional programming here just makes the code beautiful but doesn't provide any other added advantage. Although, I do understand that breaking down code into smaller functions for the sake of single responsibility and for better testability is good, is it really applicable in this case? 
Please correct me if I'm wrong. I won't make this change until you guys convince me that you have solid reasoning behind your code!